resources/type-system/flow-analysis.md

```diff
@@ -414,39 +414,67 @@ Policy:
     - and `T <: S` or (`S` is `X extends R` and `T <: R`) or (`S` is `X & R` and
       `T <: R`)
 
-  - We say that a variable `x` is promotable via assignment of an expression of
-    type `T` given variable model `VM` if
-    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
-    - and `captured` is false
-    - and `S` is the current type of `x` in `VM`
-    - and `T <: S` and not `S <: T`
-    - and `T` is a type of interest for `x` in `tested`
-
-  - We say that a variable `x` is demotable via assignment of an expression of
-    type `T` given variable model `VM` if
-    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
-    - and `captured` is false
-    - and [...promoted, declared] contains a type `S` such that `T` is `S` or
-      `T` is **NonNull(`S`)**.
-
 Definitions:
 
+- `currentType(declared, promoted)`, where `declared` is a type and `promoted`
+  is a list of types, is the type `T` , defined as follows:
+  - If `promoted` is `[]`, `T` is `declared`.
+  - Otherwise, `T` is the last type in the `promoted` list.
+
+- `demote(promoted, written)`, where `promoted` is a list of types and `written`
+  is a type, is a list obtained by deleting any elements from `promoted` that do
+  not satisfy `written <: T`. _In effect, this removes any type promotions that
+  are no longer valid after the assignment of a value of type `written`._
+
+- `toi_promote(declared, demoted, tested, written)`, where `demoted` and
+  `tested` are lists of types, and `declared` and `written` are types, is the
+  list `promoted`, defined as follows. _("toi" stands for "type of interest".)_
+  - Let `current = currentType(declared, demoted)`. _(This is the type of the
+    variable after demotions, but before type of interest promotion.)_
+  - If `written` and `current` are the same type, then `promoted` is
+    `demoted`. _(No type of interest promotion is necessary in this case.)_
```

Comment:
If `demoted` is renamed as `promotionChain` then `promoted` could be renamed as `newPromotionChain`.

Reply:
> If `demoted` is renamed as `promotionChain` then `promoted` could be renamed as `newPromotionChain`.

Done.